### PR TITLE
IPFS: use ipfs binary instead of library

### PR DESCRIPTION
**Pull request by ktock**
Fixes: https://github.com/containerd/nerdctl/issues/1578

This commit reduces dependencies for ipfs library by using ipfs binary.
`nerdctl ipfs registry up` and `nerdctl ipfs registry down` now uses systemd for managing `nerdctl ipfs registry serve` because it now depends on ipfs command available on the host rootfs.

- TODO
  - agreement on design
  - ~~update examples~~ will be done after we release the next version
  - ~~remove `github.com/ipfs/go-cid v0.3.2` (CID parser) dependency (following up PR?)~~
    - ~~We need to reimplement the parser to remove that dependency. ~~

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
Thanks, you can retain go-cid and call it a day

File: cmd/nerdctl/ipfs_linux_test.go [hunk 1]
```diff
@@ -134,13 +133,11 @@ func pushImageToIPFS(t *testing.T, base *testutil.Base, name string, opts ...str
 
 func cidOf(t *testing.T, lines []string) string {
 	assert.Equal(t, len(lines) >= 2, true)
-	c, err := cid.Decode(lines[len(lines)-2])
-	assert.NilError(t, err)
-	return "ipfs://" + c.String()
+	return "ipfs://" + lines[len(lines)-2]
 }
 
 func requiresIPFS(t *testing.T) {
```
Comment (djdongjin) [inline cmd/nerdctl/ipfs_linux_test.go:139]: maybe put in `testutil`, similar to other `RequireXXXX`.

https://github.com/containerd/nerdctl/blob/e2d618f283372f19c4fc52b809d6f6e1944ef8d3/pkg/testutil/testutil.go

File: docs/ipfs.md [hunk 2]
```diff
@@ -159,14 +167,19 @@ When you specify `--ipfs` option to `nerdctl build`, it automatically starts the
 By default, nerdctl exposes the registry at `localhost:5050`.
 You can change the address and can manually restart the registry using `nerdctl ipfs registry up` and `nerdctl ipfs registry down`.
 
+To make these commands work correctly, you need to install `nerdctl ipfs registry serve` command as a systemd service named `nerdctl-ipfs-registry-serve`.
```
Comment (djdongjin) [inline docs/ipfs.md:170]: Maybe we can have a sample `systemd service` file that can be used directly by users?

File: Dockerfile.d/test-integration-nerdctl-ipfs-registry-serve.service [hunk 3]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=nerdctl ipfs registry for integration test
+
+[Service]
+EnvironmentFile=/run/nerdctl-ipfs-registry-serve/args
```
Comment (AkihiroSuda) [inline Dockerfile.d/test-integration-nerdctl-ipfs-registry-serve.service:5]: s/args/env/ maybe

**Comment by AkihiroSuda** [comment]
> serve

server, service, or just drop it

File: cmd/nerdctl/ipfs_registry_up.go [hunk 4]
```diff
@@ -36,8 +37,17 @@ const ipfsRegistryContainerName = "ipfs-registry"
...
+		Short: "start or restart registry as a systemd service, backed by the current user's IPFS API",
+		Long: `start or restart registry as a systemd service, backed by the current user's IPFS API
+
+This command requires "nerdctl ipfs registry serve" command is registered as a systemd service named "nerdctl-ipfs-registry-serve".
+In rootless mode, use "containerd-rootless-setuptool.sh install-nerdctl-ipfs-registry-serve" to install it.
+
+Instead of using this command, you can also directly launch the registry using "nerdctl ipfs registry serve".
+
+Use "nerdctl ipfs registry down" to stop the registry service launced by "nerdctl ipfs registry up".
+`,
```
Comment (AkihiroSuda) [inline cmd/nerdctl/ipfs_registry_up.go:50]: Maybe we can just drop the up/down command and let the user install the systemd unit by themselves

File: docs/ipfs.md [hunk 5]
```diff
@@ -124,10 +132,17 @@ FROM localhost:5050/ipfs/bafkreicq4dg6nkef5ju422ptedcwfz6kcvpvvhuqeykfrwq5krazf3
...
+This allows `nerdctl build` to pull images from IPFS.
+
+```
+$ nerdctl ipfs registry serve &
+```
+
+Then you can build this Dockerfile using `nerdctl build`.
 
 ```console
-> nerdctl build --ipfs -t hello .
```
Comment (AkihiroSuda) [inline docs/ipfs.md:130]: The document may note that this flag (and up/down) is being removed in nerdctl v1.2.0.

File: cmd/nerdctl/build.go [hunk 6]
```diff
@@ -70,7 +70,7 @@ If Dockerfile is not present and -f is not specified, it will look for Container
 	buildCommand.RegisterFlagCompletionFunc("platform", shellCompletePlatforms)
 	// #endregion
 
-	buildCommand.Flags().Bool("ipfs", false, "Allow pulling base images from IPFS")
+	// buildCommand.Flags().Bool("ipfs", false, "Allow pulling base images from IPFS")
```
Comment (AkihiroSuda) [inline cmd/nerdctl/build.go:73]: Remove

**Review by AkihiroSuda — APPROVED** [review]

**Comment by AkihiroSuda** [comment]
Still draft?

**Comment by ktock** [comment]
Thank you for the review.

File: pkg/testutil/testutil.go [hunk 7]
```diff
@@ -604,6 +604,12 @@ func RequireExecutable(t testing.TB, name string) {
 	}
 }
 
+func RequireIPFS(t *testing.T) {
```
Comment (djdongjin) [inline pkg/testutil/testutil.go:607]: nit: maybe this function can be replaced by just using `RequireExecutable("ipfs")` above.

**Comment by ktock** [comment]
It looks like the ipfs registry sometimes breaks. I'll fix it.

```
WARN[0215] ipfs command failed                           error="exit status 1" stderr="Error: lock /home/rootless/.ipfs/repo.lock: someone else has the lock\n"
WARN[0216] ipfs command failed                           error="exit status 1" stderr="Error: lock /home/rootless/.ipfs/repo.lock: someone else has the lock\n"
WARN[0217] ipfs command failed                           error="exit status 1" stderr="Error: lock /home/rootless/.ipfs/repo.lock: someone else has the lock\n"
WARN[0218] ipfs command failed                           error="exit status 1" stderr="Error: lock /home/rootless/.ipfs/repo.lock: someone else has the lock\n"
WARN[0219] ipfs command failed                           error="exit status 1" stderr="Error: lock /home/rootless/.ipfs/repo.lock: someone else has the lock\n"
```

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**